SG14/inplace_function.h

```diff
@@ -29,340 +29,331 @@
+
+    static_assert(
+      Alignment % std::alignment_of<C>::value == 0,
+      "Incompatible function closure alignment"
+    );
+
+    static const vtable_t vt{detail::wrapper<C>{}};
+    vtable_ptr_ = std::addressof(vt);
+
+    new (std::addressof(storage_)) C{std::forward<C>(closure)};
+  }
+
+  inplace_function(std::nullptr_t) noexcept :
+    vtable_ptr_{std::addressof(detail::empty_vtable<R, Args...>)}
+  {}
+
+  inplace_function(const inplace_function& other) :
+    vtable_ptr_{other.vtable_ptr_}
+  {
+    vtable_ptr_->copy_ptr(
+      std::addressof(storage_),
+      std::addressof(other.storage_)
+    );
+  }
+
+  inplace_function(inplace_function&& other) :
+    vtable_ptr_{other.vtable_ptr_}
+  {
+    vtable_ptr_->move_ptr(
+      std::addressof(storage_),
+      std::addressof(other.storage_)
+    );
+  }
+
+  inplace_function& operator= (std::nullptr_t) noexcept
+  {
+    vtable_ptr_->destructor_ptr(std::addressof(storage_));
+    vtable_ptr_ = std::addressof(detail::empty_vtable<R, Args...>);
+    return *this;
+  }
+
+  inplace_function& operator= (const inplace_function& other)
+  {
+    if(this != std::addressof(other))
+    {
+      vtable_ptr_->destructor_ptr(std::addressof(storage_));
+
+      vtable_ptr_ = other.vtable_ptr_;
+      vtable_ptr_->copy_ptr(
+        std::addressof(storage_),
+        std::addressof(other.storage_)
+      );
+    }
+    return *this;
+  }
+
+  inplace_function& operator= (inplace_function&& other)
+  {
+    if(this != std::addressof(other))
+    {
+      vtable_ptr_->destructor_ptr(std::addressof(storage_));
+
+      vtable_ptr_ = other.vtable_ptr_;
+      vtable_ptr_->move_ptr(
+        std::addressof(storage_),
+        std::addressof(other.storage_)
+      );
+    }
+    return *this;
+  }
+
+  ~inplace_function()
+  {
+    vtable_ptr_->destructor_ptr(std::addressof(storage_));
+  }
+
+  R operator() (Args... args) const
+  {
+    return vtable_ptr_->invoke_ptr(
+      std::addressof(storage_),
+      std::forward<Args>(args)...
+    );
+  }
+
+  constexpr bool operator== (std::nullptr_t) const noexcept
+  {
+    return !operator bool();
+  }
+
+  constexpr bool operator!= (std::nullptr_t) const noexcept
+  {
+    return operator bool();
+  }
+
+  explicit constexpr operator bool() const noexcept
+  {
+    return vtable_ptr_ != std::addressof(detail::empty_vtable<R, Args...>);
+  }
+
+  template<size_t Cap, size_t Align>
+  operator inplace_function<R(Args...), Cap, Align>() const&
+  {
+    static_assert(detail::is_valid_inplace_dst<
+      Cap, Align, Capacity, Alignment
+    >::value);
+
+    return {vtable_ptr_, std::addressof(storage_)};
+  }
+
+  template<size_t Cap, size_t Align>
+  operator inplace_function<R(Args...), Cap, Align>() &&
+  {
+    static_assert(detail::is_valid_inplace_dst<
+      Cap, Align, Capacity, Alignment
+    >::value);
+
+    return {std::addressof(vtable_ptr_), std::addressof(storage_)};
```

Comment:
The semantic difference between line 301's

    return {std::addressof(vtable_ptr_), std::addressof(storage_)};

and line 291's

    return {vtable_ptr_, std::addressof(storage_)};

is so subtle that I missed it on first reading. I haven't really pondered whether there's a philosophical reason to express "copy" by "copy of the vptr" and "move" by "dereference of the vptr". Might I suggest that you use something like `return inplace_function<R(Args...), Cap, Align>(vtable_ptr_, std::addressof(storage_), std::true_type{})` or something, i.e., overload on a separate tag type to communicate the "should-move-from" bit, instead of overloading on the pointerness of the first argument?

Orthogonally: is there a specific reason that these two are expressed as conversion operators *to* the wider type, as opposed to constructors *from* the narrower type? Personally I rarely use conversion operators, myself, so I'm not necessarily a great judge of when it's appropriate to use them.

Reply:
On the conversion construction, it's neither philosophically nor idiomatically backed.
Just found a clearer implementation. When possible I try to avoid parameter type tags.
I've seen them mess with performance due to a changed function signature. Rather express with templates if possible. Plus it puts the type checking on the implementation side, expecting the user of your library to give you the matching tag is brittle.

I opted into conversion function over separate copy ctor, move ctor, copy assignment and move assignment, as it forced repetition or forced shared code. This solution minimizes both.
Also there is the issue of disabling the more specialized generic constructor if it's a `inplace_function` derivative. Which I found doable but ugly if done via the additional constructors and assignments.

Reply:
> expecting the user of your library to give you the matching tag is brittle

Nit: I don't understand this part, since the "user" in this case is just another part of the library class. I wasn't suggesting to expose *public* constructors taking tag types! But whichever; I can understand the rationale about avoiding extra tag-type parameters in signatures, even if I personally would expect that overhead to get inlined away.

On the conversion functions: Ah, makes sense. I hadn't thought about the amount of duplication w.r.t. "copy ctor, move ctor, copy assignment and move assignment."